src/BoundsInference.cpp

```diff
@@ -188,32 +189,82 @@ class BoundsInference : public IRMutator {
                 const Definition &s_def = specializations[i-1].definition;
 
                 // Else case (i.e. specialization condition is false)
-                for (CondValue &i : result) {
+                for (CondValue &i : args) {
+                    i.cond = simplify(!s_cond && i.cond);
+                }
+                for (CondValue &i : values) {
                     i.cond = simplify(!s_cond && i.cond);
                 }
 
                 // Then case (i.e. specialization condition is true)
-                vector<CondValue> s_result = compute_exprs_helper(s_def, is_update);
-                for (CondValue &i : s_result) {
+                vector<CondValue> s_args, s_values;
+                std::tie(s_args, s_values) = compute_exprs_helper(s_def, is_update);
+                for (CondValue &i : s_args) {
+                    i.cond = simplify(s_cond && i.cond);
+                }
+                for (CondValue &i : s_values) {
                     i.cond = simplify(s_cond && i.cond);
                 }
-                result.insert(result.end(), s_result.begin(), s_result.end());
+                values.insert(values.end(), s_values.begin(), s_values.end());
+                args.insert(args.end(), s_args.begin(), s_args.end());
+            }
+
+            // Optimization: If the args/values across specializations including
+            // the default case, are the same, we can combine those args/values
+            // into one arg/value with a const_true() condition for the purpose
+            // of bounds inference.
+            if (values.size() > 1) {
+                bool all_equal = true;
+                Expr val = values[0].value;
+                for (size_t i = 1; i < values.size(); ++i) {
+                    if (!is_one(simplify(val == values[i].value))) {
```

Comment:
I think equal(Expr, Expr) is sufficient here. They've both been simplified already in simplify specializations